app/uk/gov/hmrc/economiccrimelevyregistration/controllers/RegistrationSubmissionController.scala

```diff
@@ -16,129 +16,118 @@
-                              nrsService.submitToNrs(
-                                registration.base64EncodedFields.flatMap(_.nrsSubmissionHtml),
-                                eclRef,
-                                appConfig.eclAmendRegistrationNotableEvent
-                              )
-                            }
-                            auditService
-                              .successfulSubscriptionAndEnrolment(
-                                registration,
-                                eclRef,
-                                additionalInfo.liabilityYear
-                              )
-                            Future.successful(Ok(Json.toJson(response)))
-                          case Left(e)         =>
-                            logger.error(
-                              s"Failed to submit PDF to DMS: ${e.getMessage()}"
-                            )
-                            Future.successful(InternalServerError("Could not send PDF to DMS queue"))
+    (for {
+      registration   <- registrationService.getRegistration(id).asResponseError
+      additionalInfo <- registrationAdditionalInfoService.get(registration.internalId).asResponseError
+      _              <- (if (registration.isRegistration) {
+                           registrationValidation(registration)
+                         } else {
+                           subscriptionValidation(registration, additionalInfo)
+                         }).asResponseError
+      response       <- if (registration.isRegistration) {
+                          registerForEcl(registration, additionalInfo.liabilityYear)
+                        } else {
+                          subscribeToEcl(registration, additionalInfo)
                         }
-                    case None         =>
-                      logger.error(
-                        s"Expected eclReference to be present in additional information for amendment with internal id: ${registration.internalId}"
-                      )
+    } yield response).convertToResult(OK)
+  }
 
-                      Future.successful(
-                        InternalServerError("Expected eclReference to be present in additional information")
-                      )
+  private def registrationValidation(registration: Registration): EitherT[Future, DataValidationError, Registration] =
+    EitherT {
+      Future.successful(
+        registrationValidationService.validateRegistration(registration) match {
+          case Left(error)   => Left(error)
+          case Right(result) => Right(result)
+        }
+      )
+    }
+
+  private def registerForEcl(registration: Registration, liabilityYear: Option[Int])(implicit
+    hc: HeaderCarrier,
+    request: AuthorisedRequest[_]
+  ): EitherT[Future, ResponseError, CreateEclSubscriptionResponsePayload] =
+    for {
+      _        <- registrationValidation(registration).asResponseError
+      response <- dmsService
```

Comment:
I think it's harder to read when you pass in values into a function that still need to be worked out

What we could do instead is set those in the for-comprehension and pass them in, purely for readability: 

```
for {
dmsSubmission = registration.base64EncodedFields.flatMap(_.dmsSubmissionHtml)
now = Instant.now().truncatedTo(ChronoUnit.SECONDS)
response <- dmsService.submitToDms(dmsSubmission, now).asResponseError
}
```

Reply:
I think that I've refactored properly but please take a look :)